crates/shared/src/price_estimation/trade_verifier.rs

```diff
@@ -74,29 +76,61 @@ impl TradeVerifier {
     pub fn new(
         simulator: Arc<dyn CodeSimulating>,
         code_fetcher: Arc<dyn CodeFetching>,
+        trusted_tokens: AutoUpdatingTokenList,
         settlement: H160,
         native_token: H160,
     ) -> Self {
         Self {
             simulator,
             code_fetcher,
+            trusted_tokens,
             settlement,
             native_token,
         }
     }
-}
 
-#[async_trait::async_trait]
-impl TradeVerifying for TradeVerifier {
-    async fn verify(
+    /// Encodes all interactions that are needed taking the `internalize` flag
+    /// into account. Returns an error if solver requested to internalize an
+    /// interaction that must not be internalized.
+    fn interactions_to_encode(
```

Comment:
Why is this logic in `shared`? Can’t it be - like for solve requests - the driver that decides which interactions to internalise? In fact I’m not even sure why we are using a different route for quotes? To me a quote is a single order solve request (with insta reveal).

Reply:
The reason is that with verified quotes we actually want to verify that the output the solver advertises is actually what the call data generates.
To do this we inject custom interactions into the execution plan that measures sell or buy token balances at critical points throughout the settlement. Of course we could inject these interactions in the driver but then the logic the injects the interactions and the logic that reads the balances is so far apart that it would just be very hard to follow.

I guess we could also make the driver return a fully formed settlement but then we'd have to inject the balance checking interactions into the encoded calldata. Which sounds doable but way more complicated and more effort than it's worth IMO.

Reply:
Theoretically you can run into edge cases with pre-interactions.
For example you could request a quote for buying 2000 USDC for USDT using a pre-interaction that unstakes 2000 USDT from an AMM.
If we compute the `out_amount` as `usdt_before_settlement - usdt_after_settlement` our math won't make sense because it would look like we bought 2000 USDC for 0 USDT.
With our current logic we would check the USDT balance after running every pre-interation and before running any post-interactions (those don't matter for quoting anyway). This allows us to see that we actually spent the 2000 USDT we unstaked with the pre-interaction.

Reply:
If post interactions are not an issue for quoting we could probably accomplish the same with a second simulation that only executes the pre interaction and measures the sell token balance before. And if the pre interaction somehow interferes with the proposed execution we'll still notice during the complete execution. Leaving just the trade preparations (e.g. approval, wrapping) to do in the API.